File: cli-proxy-logger/src/sse.js
// Incremental Server-Sent Events parser.
//
// Feed it raw decoded text chunks via push(); it emits one object per complete
// SSE event: { event: string|null, data: string, raw: string }.
// Data lines are joined with "\n" per the SSE spec. The proxy uses this on a
// COPY of the upstream stream — the original bytes are always forwarded to the
// client untouched, so a parse error here can never corrupt the CLI's stream.

export class SSEParser {
  constructor() {
    this._buf = '';
    this._listeners = [];
  }

  on(_evt, cb) {
    this._listeners.push(cb);
    return this;
  }

  _emit(evt) {
    for (const cb of this._listeners) {
      try {
        cb(evt);
      } catch {
        // ignore listener errors; parsing must never break forwarding
      }
    }
  }

  push(text) {
    // Append the new text and peel off complete events. Because the network
    // delivers arbitrary byte boundaries, an event may be split across two
    // push() calls — so we always retain the trailing, possibly incomplete,
    // fragment in this._buf for next time. That is the essence of "incremental"
    // parsing: never assume a chunk ends on an event boundary.
    this._buf += text;
    // SSE events are separated by a blank line. Handle \n\n and \r\n\r\n.
    const normalized = this._buf.replace(/\r\n/g, '\n');
    const parts = normalized.split('\n\n');
    // The last element is whatever came after the final blank line; it may be
    // a partial event, so put it back in the buffer.
    this._buf = parts.pop() ?? '';
    for (const block of parts) {
      if (block.trim() === '') continue;
      this._emit(this._parseBlock(block));
    }
  }

  flush() {
    if (this._buf.trim() !== '') {
      this._emit(this._parseBlock(this._buf.replace(/\r\n/g, '\n')));
    }
    this._buf = '';
  }

  _parseBlock(block) {
    let event = null;
    const dataLines = [];
    for (const line of block.split('\n')) {
      if (line.startsWith(':')) continue; // comment
      const idx = line.indexOf(':');
      const field = idx === -1 ? line : line.slice(0, idx);
      let value = idx === -1 ? '' : line.slice(idx + 1);
      if (value.startsWith(' ')) value = value.slice(1);
      if (field === 'event') event = value;
      else if (field === 'data') dataLines.push(value);
    }
    return { event, data: dataLines.join('\n'), raw: block };
  }
}

File: cli-proxy-logger-py/cli_proxy_logger/sse.py
"""Incremental Server-Sent Events parser.

Feed it raw decoded text chunks via ``push()``; it invokes the registered
listener once per complete SSE event with a dict ``{event, data, raw}``. Data
lines are joined with "\\n" per the SSE spec. The proxy uses this on a COPY of
the upstream stream -- the original bytes are always forwarded to the client
untouched, so a parse error here can never corrupt the CLI's stream.
"""


class SSEParser:
    def __init__(self):
        self._buf = ""
        self._listeners = []

    def on(self, _evt, cb):
        self._listeners.append(cb)
        return self

    def _emit(self, evt):
        for cb in self._listeners:
            try:
                cb(evt)
            except Exception:
                # ignore listener errors; parsing must never break forwarding
                pass

    def push(self, text):
        # Append the new text and try to peel off complete events. Because the
        # network delivers arbitrary byte boundaries, an event may be split
        # across two push() calls -- so we always retain the trailing, possibly
        # incomplete, fragment in self._buf for next time. This is the essence
        # of "incremental" parsing: never assume a chunk ends on an event
        # boundary.
        self._buf += text
        # SSE events are separated by a blank line. Handle \n\n and \r\n\r\n.
        normalized = self._buf.replace("\r\n", "\n")
        parts = normalized.split("\n\n")
        # The last element is whatever came after the final blank line; it may
        # be a partial event, so put it back in the buffer.
        self._buf = parts.pop() if parts else ""
        for block in parts:
            if block.strip() == "":
                continue
            self._emit(self._parse_block(block))

    def flush(self):
        if self._buf.strip() != "":
            self._emit(self._parse_block(self._buf.replace("\r\n", "\n")))
        self._buf = ""

    @staticmethod
    def _parse_block(block):
        event = None
        data_lines = []
        for line in block.split("\n"):
            if line.startswith(":"):
                continue  # comment
            idx = line.find(":")
            field = line if idx == -1 else line[:idx]
            value = "" if idx == -1 else line[idx + 1:]
            if value.startswith(" "):
                value = value[1:]
            if field == "event":
                event = value
            elif field == "data":
                data_lines.append(value)
        return {"event": event, "data": "\n".join(data_lines), "raw": block}
